### ISSUE TITLE
Issue decoding Parquet file with a column ending with a null page

### ISSUE THREAD

**Issue by mdrach** [jorgecarleitao/parquet2#79]
When trying to decompress a numeric column in a parquet file that ends with a page of nulls, parquet2 panics.  Here is a test case that reproduces the conditions of the failure.

https://github.com/jorgecarleitao/parquet2/commit/85a092f68326da1d555742f59e6aac2caf91b53e

The backtrace of the error is here:
```
---- stage::tests::retrieve_s3_parquet_test stdout ----
thread 'tokio-runtime-worker' panicked at 'called `Option::unwrap()` on a `None` value', /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/parquet2-0.9.0/src/encoding/hybrid_rle/mod.rs:41:32
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:100:14
   2: core::panicking::panic
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:50:5
   3: core::option::Option<T>::unwrap
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/option.rs:746:21
   4: parquet2::encoding::hybrid_rle::read_next
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/parquet2-0.9.0/src/encoding/hybrid_rle/mod.rs:41:17
   5: parquet2::encoding::hybrid_rle::HybridRleDecoder::new
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/parquet2-0.9.0/src/encoding/hybrid_rle/mod.rs:63:21
   6: arrow2::io::parquet::read::primitive::basic::read_dict_buffer_optional
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/arrow2-0.9.0/src/io/parquet/read/primitive/basic.rs:38:9
   7: arrow2::io::parquet::read::primitive::basic::extend_from_page
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/arrow2-0.9.0/src/io/parquet/read/primitive/basic.rs:184:13
   8: arrow2::io::parquet::read::primitive::iter_to_array
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/arrow2-0.9.0/src/io/parquet/read/primitive/mod.rs:86:13
   9: arrow2::io::parquet::read::page_iter_to_array
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/arrow2-0.9.0/src/io/parquet/read/mod.rs:359:20
  10: arrow2::io::parquet::read::column_iter_to_array
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/arrow2-0.9.0/src/io/parquet/read/mod.rs:438:25
  11: ingest::stage::load_parquet_row_group::{{closure}}::{{closure}}::{{closure}}
             at ./src/stage.rs:404:17
  12: tokio::runtime::enter::exit
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/enter.rs:90:9
  13: tokio::runtime::thread_pool::worker::block_in_place
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/thread_pool/worker.rs:332:9
  14: tokio::task::blocking::block_in_place
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/task/blocking.rs:77:9
  15: ingest::stage::load_parquet_row_group::{{closure}}::{{closure}}
             at ./src/stage.rs:396:37
  16: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  17: ingest::stage::load_parquet_row_group::{{closure}}
             at ./src/stage.rs:357:1
  18: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  19: ingest::stage::load_parquet_shard::{{closure}}::{{closure}}
             at ./src/stage.rs:335:25
  20: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  21: ingest::stage::load_parquet_shard::{{closure}}
             at ./src/stage.rs:311:1
  22: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  23: ingest::stage::load_shard::{{closure}}::{{closure}}
             at ./src/stage.rs:156:17
  24: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  25: ingest::stage::load_shard::{{closure}}
             at ./src/stage.rs:116:1
  26: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  27: <sentry_core::futures::SentryFuture<F> as core::future::future::Future>::poll::{{closure}}
             at /home/ubuntu/.cargo/git/checkouts/sentry-rust-523ec8d3fcd6cdc5/2a15988/sentry-core/src/futures.rs:40:30
  28: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panic/unwind_safe.rs:271:9
  29: std::panicking::try::do_call
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:403:40
  30: __rust_try
  31: std::panicking::try
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:367:19
  32: std::panic::catch_unwind
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panic.rs:133:14
  33: sentry_core::hub::Hub::run
             at /home/ubuntu/.cargo/git/checkouts/sentry-rust-523ec8d3fcd6cdc5/2a15988/sentry-core/src/hub.rs:229:26
  34: <sentry_core::futures::SentryFuture<F> as core::future::future::Future>::poll
             at /home/ubuntu/.cargo/git/checkouts/sentry-rust-523ec8d3fcd6cdc5/2a15988/sentry-core/src/futures.rs:40:13
  35: <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/task/task_local.rs:280:28
  36: tokio::task::task_local::TaskLocalFuture<T,F>::with_task
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/task/task_local.rs:272:9
  37: <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/task/task_local.rs:280:9
  38: service::context::Context::scope::{{closure}}
             at /home/ubuntu/core/service/src/context.rs:111:9
  39: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  40: service::context::InContext::in_current_context::{{closure}}
             at /home/ubuntu/core/service/src/context.rs:364:9
  41: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  42: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/future.rs:119:9
  43: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.29/src/instrument.rs:272:9
  44: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/task/core.rs:161:17
  45: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/loom/std/unsafe_cell.rs:14:9
  46: tokio::runtime::task::core::CoreStage<T>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/task/core.rs:151:13
  47: tokio::runtime::task::harness::poll_future::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/task/harness.rs:461:19
  48: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panic/unwind_safe.rs:271:9
  49: std::panicking::try::do_call
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:403:40
  50: __rust_try
  51: std::panicking::try
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:367:19
  52: std::panic::catch_unwind
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panic.rs:133:14
  53: tokio::runtime::task::harness::poll_future
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/task/harness.rs:449:18
  54: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/task/harness.rs:98:27
  55: tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/task/harness.rs:53:15
  56: tokio::runtime::task::raw::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/task/raw.rs:113:5
  57: tokio::runtime::task::raw::RawTask::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/task/raw.rs:70:18
  58: tokio::runtime::task::LocalNotified<S>::run
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/task/mod.rs:343:9
  59: tokio::runtime::thread_pool::worker::Context::run_task::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/thread_pool/worker.rs:420:13
  60: tokio::coop::with_budget::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/coop.rs:106:9
  61: std::thread::local::LocalKey<T>::try_with
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/thread/local.rs:399:16
  62: std::thread::local::LocalKey<T>::with
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/thread/local.rs:375:9
  63: tokio::coop::with_budget
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/coop.rs:99:5
  64: tokio::coop::budget
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/coop.rs:76:5
  65: tokio::runtime::thread_pool::worker::Context::run_task
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/thread_pool/worker.rs:419:9
  66: tokio::runtime::thread_pool::worker::Context::run
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/thread_pool/worker.rs:386:24
  67: tokio::runtime::thread_pool::worker::run::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/thread_pool/worker.rs:371:17
  68: tokio::macros::scoped_tls::ScopedKey<T>::set
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/macros/scoped_tls.rs:61:9
  69: tokio::runtime::thread_pool::worker::run
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/thread_pool/worker.rs:368:5
  70: tokio::runtime::thread_pool::worker::block_in_place::{{closure}}::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/thread_pool/worker.rs:324:41
  71: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/blocking/task.rs:42:21
  72: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/task/core.rs:161:17
  73: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/loom/std/unsafe_cell.rs:14:9
  74: tokio::runtime::task::core::CoreStage<T>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/task/core.rs:151:13
  75: tokio::runtime::task::harness::poll_future::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/task/harness.rs:461:19
  76: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panic/unwind_safe.rs:271:9
  77: std::panicking::try::do_call
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:403:40
  78: __rust_try
  79: std::panicking::try
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:367:19
  80: std::panic::catch_unwind
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panic.rs:133:14
  81: tokio::runtime::task::harness::poll_future
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/task/harness.rs:449:18
  82: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/task/harness.rs:98:27
  83: tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/task/harness.rs:53:15
  84: tokio::runtime::task::raw::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/task/raw.rs:113:5
  85: tokio::runtime::task::raw::RawTask::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/task/raw.rs:70:18
  86: tokio::runtime::task::UnownedTask<S>::run
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/task/mod.rs:379:9
  87: tokio::runtime::blocking::pool::Inner::run
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/blocking/pool.rs:264:17
  88: tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/blocking/pool.rs:244:17
```

The problematic file is sensitive, but the it's a mostly-null numeric column.  The last page contains ~300 null values.

**Comment by mdrach** [comment]
Happy to help fix this as this as our use-case is a bit time-sensitive, but would love some guidance on how to approach this if possible.

**Comment by jorgecarleitao** [comment]
Sorry to hear that  and thanks a lot for the offer!

Would it be possible to inspect the last pages' definition levels? We should trigger the error when reading it with `parquet2::encoding::hybrid_rle::Decoder` with a bit width of 1. If they are all nulls, the bytes in the def levels should be small. From here, we should be able to tackle the issue in parquet2, since we know what the expected result is.

One way to read the last page's encoded: read the last page, decompress it, get the pages' buffer, split it in def and values (e.g. `parquet2::page::split_buffer(&page, descriptor).1`)

**Comment by mdrach** [comment]
Thanks for the quick reply!  The issue happens during decompression, so afaict it's not possible to call split_buffer from my test code.  I'm trying to find another way of getting this information.

**Comment by jorgecarleitao** [comment]
No worries :) from the stack trace it happens on deserialization (!= decopression).

Read -> CompressedDataPage -> DecompressedDataPage -> array. 

We have an issue on the third arrow where we deserialize the def level buffer into the validity. Arrow2 only sees deckmpressed pages :)

**Comment by mdrach** [comment]
Ah, gotcha.   Here is the result of split_buffer:

```
rep levels: []
def levels: [172, 5, 0]
encoded values: [1]
```

**Comment by jorgecarleitao** [comment]
Thanks, that was helpful! I've identified and fixed the issue in #80. It has been released to crates on 0.9.2.
